### PR TITLE
feat(lesson): pep8, linters & code formatters

### PR DESCRIPTION
**Pull request by lwasser**
todo: add to bottom  of intro clean code page:

`[^zenofpython]: Paste this into your shell: `python -c 'import this'``

### CODE REVIEW

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [inline clean-modular-code/python-pep-8.md:303]: "that are included with python," "that you don't have to install," with link to https://docs.python.org/3/library/index.html ?

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [inline clean-modular-code/python-pep-8.md:322]: add the third block of 

```python
# imports from your local package
import mypackage
from mypackage.internal import _secret_function
```

Comment (sneakers-the-rat) [reply clean-modular-code/python-pep-8.md:322]: oh i see you have a separate section for that. my b. i just figured showing all three of them in the example that follows the list of 3 would be nice, but this is a draft and i'm jumping the gun on giving comments lol

Comment (lwasser) [reply clean-modular-code/python-pep-8.md:322]: go for it!! comment away. i just combined two pages. i'm totally fine with consolidating!!  

 This is good, too, because in our second workshop, they will create a package and import it. so they will get to try this out!!

Comment (lwasser) [reply clean-modular-code/python-pep-8.md:284]: I've just pushed a new version of this and made all of the fixes that you suggested. I also redesigned the structure to be:

* Example of good vs not as good code format. 
* How to apply tools to help you make your code prettier :) 
* Why use these tools and standards
* Some rules to remember 

That way, we start with something that doesn't have to be so hard and end with the details that many people will skip (I think?).

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:177]:
```suggestion
There is a style guide devoted to Python PEP 8 standards that you can read [here](https://www.python.org/dev/peps/pep-0008/#naming-conventions). However, below are a handful of PEP 8 Rules that you can consider following when writing code. 
```

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:196]:
```suggestion
# Call to create an object of a class
data_processor = DataProcessor()
numbers = [10, 20, 30, 40, 50]
# Examples of variable names and method call
average_value = data_processor.calculate_average(numbers)
```

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:197]:
```suggestion
# Example of function call
```

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:211]:
```suggestion
* White Space: Add space between sections of code to improve readability.
```

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:213]:
```suggestion
* Add a blank line before a single-line comment (unless it is the first line of a cell in Jupyter Notebook or part of a code block)
```

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:239]: This is more accurate.

```suggestion
For the entire pep-8 style guide see: <a href="https://peps.python.org/pep-0008/" target="_blank">Python's PEP 8 Style Guide</a>.
```

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:280]: Let's alphabetize these imports.

File: clean-modular-code/python-pep-8.md
Comment (willingc) [inline clean-modular-code/python-pep-8.md:292]:
```suggestion
Organizing your imports this way ensures your code is readable and follows widely accepted practices. Importing libraries at the top also makes it easier to scan for a specific library, debug dependencies, and see which dependencies are required to run the code.
```

**Review by willingc — COMMENTED** [review]
Nice!

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [inline clean-modular-code/python-pep-8.md:223]: an idea: all the whitespace stuff could go in a single example like the capitalization stuff

```python
from itertools import counter


# two lines above a function definition
def my_function():
    count = counter()
    values = [next(count) for i in range(10)]

    # one line to break up code into related chunks
    plt.plot(values)


# Two lines before a class definition
class MyClass:
    
    # one line before a method definition
    def my_method(): ...
```

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [inline clean-modular-code/python-pep-8.md:228]: it won't let me comment on line 231 but i think that should be `plt.subplots()`

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [inline clean-modular-code/python-pep-8.md:236]: this is odd to me to have as its own top-level section after listing a bunch of pep 8 rules in the previous subsection, and it seems to be the same information as the example starting on line 179? i would combine those things and put them under a section on various types of conventions. i'll give an example ToC in the main comment

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [inline clean-modular-code/python-pep-8.md:246]: section starts "first" but there is no next!

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [reply clean-modular-code/python-pep-8.md:284]: I hear you on that but i think that currently it's a little bewildering the other way where we're missing some introductory information on what we're talking about and why first, about to post the review

edit: oh wait it makes this comment part of the review even when we started it before that lol my b

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [reply clean-modular-code/python-pep-8.md:280]: isort makes it a little hard to know what the rules are exactly but if we want to add the sorting rules in here, here are the most common rules (in addition to grouping and alphabetization):

- modules:
  - module-only imports (`import x`) come before imports with objects (`from x import y`)
  - within that, import statements sorted alphabetically by module name
- objects within a module import (the `y` in `from x import y`)
  - first sort by whether the first letter is capitalized or not
  - then alphabetically

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [inline clean-modular-code/python-pep-8.md:292]: and merge conflicts, but i think basically we should consolidate "why code style" at the beginning because it's mostly the same reasons

**Review by sneakers-the-rat — APPROVED** [review]
OK Leah, love it, and i hadn't seen the rest of this so i got carried away reading around the rest and i love that too. I think my role in reviewing your writing is always just "take all the really good stuff here and suggest totally switching the order around," which is hopefully clear at this point that you can always take or leave, totally up to you. I have been going through this recently with some new labbies who are just getting booted into collaborative coding for the first time so this is very top of mind for me and wished i had a resource like this!

One thing i'm missing here is an overview of "what are the kinds of things that might be subject to style rules." The examples are good, but some bigger picture categories that i could group things into: naming, comments, whitespace, character use (tabs/spaces, '/"), syntax (eg. using a ternary rather than if/else, etc.). I think reordering that section, adding subheadings between each of them, would make it easier to digest :)

Current structure:

```
- Why code style and Python PEP 8 matters
- How to Apply PEP 8 Code Style Standards
  - For .py Files
  - For Jupyter Notebooks
  - Running These Tools
- Why use code standards when writing Python code
  - Some PEP 8 rules to remember
- PEP 8 naming conventions
- Best practices for importing libraries
  - Import Python libraries at the top of your code
  - Organize your imports into groups
  - Why organize imports?  
```

So this is how i would do it (but as always not saying you have to):

```
- What is code style
  - introductory example (the formatted/unformatted sentence & code examples) 
  - why does it matter for code (code is read more than it's written, merges, etc. all the reasons)
- Common python conventions
  - Formatting (big, top-level, "how is the document structured")
    - Whitespace
    - Imports & Order  
  - Naming (medium, single objects)
    - Capitalization
    - Special characters like l/l
  - Annotations (Notes to self!)
    - Comments
    - Docstrings  
- Tooling
  - what are linters & formatters?
  - examples
    - .py files
    - notebooks
  - pre-commit (why you would do it and links to other guides on how to do it)
  - IDE integration (links to examples)
```

One other sorta tangential thing- think it's important to give the other side so this too, that they shouldn't feel restricted by linters/formatters! they should use them if they support their work and help them think clearly. If they find a rule annoying? they should turn it off. it's their call. i don't mean to be annoying or overly ideological about this since i am [known to make a big deal about standardization in devops](https://github.com/pyOpenSci/python-package-guide/issues/317) but figured i'd bring it up.

File: clean-modular-code/python-pep-8.md
Comment (willingc) [reply clean-modular-code/python-pep-8.md:60]: `E501` is something that I always suppress as unneeded in modern code.

File: clean-modular-code/python-pep-8.md
Comment (willingc) [reply clean-modular-code/python-pep-8.md:127]: As a co-maintainer of Black, the original intent was to apply rules to minimize the size of diffs and to make code review easier. 

@sneakers-the-rat I would be interested when black and ruff contradict each other.

**Comment by willingc** [comment]
@lwasser @sneakers-the-rat Let's split this PR up or merge as is and then do another iteration. I'm finding it difficult to work with all three files in the same PR and follow proposed edits.

I would love to see something more along the lines:
- Create maintainable code
    - Why it's helpful
    - What are elements of maintainable code
        - Clean
        - Modular
        - Readable
        - Consistent code style
        - DRY: minimize code repetition
- Code style
    - What is code style
    - Why it's important
      - Easier to maintain
      - Easier to change and review
    - Types of tools (generic)
       - Linters
       - Formatters
       - Integrated into editors, IDEs, CI
- Python code style: PEP8 and beyond
   - PEP8
   - black

Etc.

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [reply clean-modular-code/python-pep-8.md:127]: That's the main one I have experienced! Mostly when writing `@override`s, black will want to put the ellipsis on the same line and ruff will want to put it on the following line, its not a big deal I just have to remember to run black after ruff which is what I think you're supposed to do anyway?

File: clean-modular-code/python-pep-8.md
Comment (sneakers-the-rat) [reply clean-modular-code/python-pep-8.md:60]: Thats reassuring to know! I usually bump it up to 90 :)

**Comment by willingc** [comment]
@lwasser Let's merge this as is and iterate on separate PRs.

File: clean-modular-code/python-pep-8.md
Comment (lwasser) [reply clean-modular-code/python-pep-8.md:284]: 😆 no worries!! i think / hope that i addressed your comment!! I am going to focus on merging this today give we are a week out BUT we can always revisit all of this content and update/ enhance / beautify, etc in the future!! 🚀 

thank you for this feedback!!

**Comment by lwasser** [comment]
Ok fantastic yall. Let's do this

1. When I'm back at.my computer I'll turn the bigger comments into issues!
2. I already committed the online edits!
3. I'll merge so we can then iterate on smaller prs. This likely will happen after the fall festival but I'm excited to work on this with everyone!! And I'll keep each pr small from here on in..

I hope that works for you both!! ✨️

File: clean-modular-code/python-pep-8.md
Comment (lwasser) [reply clean-modular-code/python-pep-8.md:64]: Oh wow—I always use that 79-character limit. It's just a habit at this point—all of my text editors are set up that way. But what is annoying is when it's applied to docstrings. I never thought about it as "optional," but I guess it is!

File: clean-modular-code/python-pep-8.md
Comment (lwasser) [reply clean-modular-code/python-pep-8.md:131]: So, I've stopped using Ruff on some projects and have migrated to Ruff for code formatting, too, for ease of tooling. But i still use black as a code formatter in vscode. It's so interesting that there are so many different options when thinking about linting and code formatting tools. I say this because only a few years ago i was using autopep8 for software and the autopep8 tool for jupyter in teaching.

**Pull request merged**
